exchange/auction_test.go

```diff
@@ -94,92 +112,151 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
+		var formattedData string
+		var compareString string
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
-			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
+		for _, cExpected := range specData.ExpectedCacheables {
+			formattedData = strings.Replace(string(cExpected.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("%s_%d", formattedData, cExpected.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
 			}
-			if !keymatch {
-				continue
+			ht[compareString].freq += 1
+			if targData.includeCacheVast {
+				ht[compareString].expectedKeys = append(ht[compareString].expectedKeys, cExpected.Key)
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+		}
+		for _, cFound := range cache.items {
+			formattedData = strings.Replace(string(cFound.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("\"%s\"_%d", formattedData, cFound.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
+			}
+			ht[compareString].freq -= 1
+			if targData.includeCacheVast {
+				ht[compareString].actualKeys = append(ht[compareString].actualKeys, cFound.Key)
+			}
+		}
+		for k, cachedElements := range ht {
+			if cachedElements.freq > 0 {
+				t.Errorf("%s:  [CACHE_ERROR] Cache inconsistensy. Element %s was not expected to get cached\n", fileDisplayName, k)
+			} else if cachedElements.freq < 0 {
+				t.Errorf("%s:  [CACHE_ERROR] Cache inconsistensy. We cached some more elements (%s) than expected.\n", fileDisplayName, k)
+			} else if targData.includeCacheVast {
+				for i := 0; i < len(cachedElements.expectedKeys); i++ {
+					found := false
+					for j := 0; j < len(cachedElements.actualKeys); j++ {
+						if strings.HasPrefix(cachedElements.expectedKeys[i], cachedElements.actualKeys[j]) {
+							found = true
+						}
+					}
+					if !found {
+						t.Errorf("[CACHE_ERROR] Key \"%s\" was expected to be cached but was not.\n", cachedElements.expectedKeys[i])
+					}
+				}
```

Comment:
I just realized that the above code confirms only that expected custom keys are found, but does not check that unexpected custom keys were found.  This can probably be done most easily by comparing length of expectedKeys array to length of actualKeys array.  I suspect you can do that before the above loop and execute the loop only if the lengths match.

Reply:
Lines 193 to 197 perform the checks you mention we only go ahead to compare the cached values in line 198 when lengths are not equal 
```191     _ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
192
193     if len(specData.ExpectedCacheables) > len(cache.items) {
194         t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
195     } else if len(specData.ExpectedCacheables) < len(cache.items) {
196         t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
197     } else { // len(specData.ExpectedCacheables) == len(cache.items)
198         var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
199         var formattedData string
200         var compareString string
201
202         for _, cExpected := range specData.ExpectedCacheables {
203             formattedData = strings.Replace(string(cExpected.Data), "\\", "", -1)
204 +-- 40 lines: formattedData = strings.Replace(formattedData, " ", "", -1)----------------------------------------------------------------------------------------------------------------------------
244                 }
245             }
246         }
247     }
248 }
```

Reply:
I'm referring to the comparisons of **expectedKeys** and **actualKeys** (which is the code that was in the loop immediately above my comment), not the examination of lengths of **specData.ExpectedCacheables** and **cache.items**.

Reply:
We are comparing them, its comparison is reflected in the 'freq' field with the difference that the freq field is better because it gives us the exact number of elements that mismatch, which we probably should mention in the comments, right? Again as stated in a response above, I agree with you that the error messages should change to:
```
	if cachedElements.freq > 0 {
		t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. We cached %d less element of %s than expected\n", fileDisplayName,cachedElements.freq, k)
	} else if cachedElements.freq < 0 {
		t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. We cached element %d more elements of %s than expected\n", fileDisplayName,cachedElements.freq, k)
	} else if targData.includeCacheVast {
		for i := 0; i < len(cachedElements.expectedKeys); i++ {
``` 
So we can have an output like shown below:
Cached values:
```
Expected      Actual
bidA           bidA
bidA           bidA
bidB           bidB
bidC           bidB
bidC
```
Hash table:
```
ht       freq    expectedKeys  actualKeys
bidA360    0          CatA       CatA
bidB220   -1          CatB       CatB
bidC360    2          CatC    
```
Output message: `We cached 1 more elements 'CatB' than we expected` . I realize that we also have a mismatch in BidC but once bidB gets sorted out, BidC will get its own error message when we run the tests again.

Reply:
There's no connection between cachedElements.freq and either expectedKeys or actualKeys.  If there were, you wouldn't need any code to compare keys when cachedElements.freq is 0.  Example:
- only expected data ht value is bid001 and it has no expected key.  When it gets added, freq is incremented to 1.
- only cached value is bid001 and it has a key.  When it is found, freq is decremented to 0.
This test should fail because we weren't expecting a key value but we got one.  Current code would miss that.

Reply:
The hash table approach has been modified for an O(n^2) approach where fields are compared directly with one another.